x/shield/keeper/proposal.go

```diff
@@ -17,3 +21,339 @@ func (k Keeper) ClaimUnlock(ctx sdk.Context, proposalID, poolID uint64, loss sdk
+func (k Keeper) SetReimbursement(ctx sdk.Context, proposalID uint64, payout types.Reimbursement) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(payout)
+	store.Set(types.GetReimbursementKey(proposalID), bz)
+}
+
+// GetReimbursement get a reimbursement in store.
+func (k Keeper) GetReimbursement(ctx sdk.Context, proposalID uint64) (types.Reimbursement, error) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.GetReimbursementKey(proposalID))
+	if bz != nil {
+		var reimbursement types.Reimbursement
+		k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &reimbursement)
+		return reimbursement, nil
+	}
+	return types.Reimbursement{}, types.ErrCompensationNotFound
+}
+
+// DeleteReimbursement deletes a reimbursement.
+func (k Keeper) DeleteReimbursement(ctx sdk.Context, proposalID uint64) error {
+	store := ctx.KVStore(k.storeKey)
+	if _, err := k.GetReimbursement(ctx, proposalID); err != nil {
+		return err
+	}
+	store.Delete(types.GetReimbursementKey(proposalID))
+	return nil
+}
+
+// CreateReimbursement creates a reimbursement.
+func (k Keeper) CreateReimbursement(ctx sdk.Context, proposalID uint64, poolID uint64, rmb sdk.Coins, beneficiary sdk.AccAddress) error {
+	totalCollateral := k.GetTotalCollateral(ctx)
+	payoutRatio := rmb.AmountOf(k.BondDenom(ctx)).ToDec().Quo(totalCollateral.ToDec())
+	purchaseRatio := k.GetTotalShield(ctx).ToDec().Quo(totalCollateral.ToDec())
+	// FIXME: Consider leftovers.
+	// FIXME: Accumulative errors could be large when the number of providers is large.
+	for _, provider := range k.GetAllProviders(ctx) {
+		payout := provider.Collateral.ToDec().Mul(payoutRatio).TruncateInt()
+		purchased := provider.Collateral.ToDec().Mul(purchaseRatio).TruncateInt()
+		if err := k.MakePayoutByProvider(ctx, provider.Address, purchased, payout); err != nil {
+			panic(err)
+		}
+	}
+	reimbursement := types.NewReimbursement(rmb, beneficiary, ctx.BlockTime().Add(k.GetClaimProposalParams(ctx).PayoutPeriod))
+	k.SetReimbursement(ctx, proposalID, reimbursement)
+	return nil
+}
+
+// PayFromDelegation reduce provider's delegations and transfer tokens to the shield module account.
+func (k Keeper) PayFromDelegation(ctx sdk.Context, providerAddr sdk.AccAddress, payout sdk.Int) {
+	delegations := k.sk.GetAllDelegatorDelegations(ctx, providerAddr)
+	totalDelAmountDec := sdk.ZeroDec()
+	for _, del := range delegations {
+		val, found := k.sk.GetValidator(ctx, del.GetValidatorAddr())
+		if !found {
+			panic("validator is not found")
+		}
+		totalDelAmountDec = totalDelAmountDec.Add(val.TokensFromShares(del.GetShares()))
+	}
+
+	payoutDec := payout.ToDec()
+	remainingDec := payoutDec
+	for i := range delegations {
+		val, found := k.sk.GetValidator(ctx, delegations[i].GetValidatorAddr())
+		if !found {
+			panic("validator is not found")
+		}
+		delAmountDec := val.TokensFromShares(delegations[i].GetShares())
+		var ubdAmountDec sdk.Dec
+		if totalDelAmountDec.GT(payoutDec) {
```

Comment:
I'm having trouble understanding this if statement. Wouldn't it always hold when we're here?

Reply:
It could be `totalDelAmount.Dec` == `payoutDec`, but you're right this condition can be removed. I added this just to avoid calculations when they are not needed.